src/object/obj_internal.h

```diff
@@ -318,7 +318,8 @@ static inline bool
 obj_retry_error(int err)
 {
 	return err == -DER_TIMEDOUT || err == -DER_STALE ||
-	       err == -DER_INPROGRESS || daos_crt_network_error(err);
+	       err == -DER_INPROGRESS || err == -DER_AGAIN ||
```

Comment:
probably with some cases the DER_AGAIN need to return to user (ask user to retry in API level rather than automatically retry in obj module), not sure how to differentiate it... (maybe define another DER_RETRY for internal try again, and keep DER_AGAIN to return to user?, not very sure though, maybe fine for now) @wangdi1 ?

Reply:
In what cases should the retry be done at the API level?

Reply:
one (cornel) case I ever discussed with @gnailzenh  recently is that for EC obj fetch, if user provides unknown record size, then client does not know how to split the sgl for different targets. Then can firstly return rec_size to user and ask user to recall the API again with the provided rec_size (by return EAGAIN?). But just now that i think seems fine to retry it in dc obj layer rather than ask user recall API for that case (Liang seems prefer to just ask user to retry/redo it to avoid adding another code path in dc obj).

I am not sure if there is other cases that need to return EAGAIN to user, just feel maybe some cases need that later.
But I am fine with your current patch, we can change it later if need to return EAGAIN to user.